### PR TITLE
clj-kondo support

### PR DESCRIPTION
**Pull request by owenRiddy**
This is maybe more for discussion as a companion to https://github.com/typedclojure/typedclojure/issues/123 - I mean something like this.

So a recap on the problem statement: When I add some typed Clojure to a random Clojure project, it becomes a mass of squiggles because clj-kondo gets stressed easily and stress leads to squiggles. We see several squiggle-reducing tactics here:

* Add a declare to `clojure.core.typed` (completely untested). clj-kondo struggles with macro-created vars.
* The folder structure is significant, in theory it might [let the config get auto-installed](https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#exporting-and-importing-configuration). (90% untested). 
* When clj-kondo sees `clojure.core.typed/defn` it rewrites it as a plain `defn` with no type hints for linting purposes. (I tested this)
* Make code like `(t/defn foo [bar :- A] bar)` a linter error. It should be `(t/defn foo [bar :- A] :- A bar)`. (untested and unjustified).

I put this together with care but also quickly and without much testing:
* I'm not sure how typed Clojure's project structure works.
* You decide that the maintenance burden is a problem or similar.

### CODE REVIEW

File: resources/clj-kondo.exports/typed/clj.checker/hooks/typedclojure.clj [hunk 1]
```diff
@@ -0,0 +1,50 @@
+(ns hooks.typed
```
Comment (owenRiddy) [inline resources/clj-kondo.exports/typed/clj.checker/hooks/typedclojure.clj:1]: `hooks.typedclojure`.

**Comment by frenchy64** [comment]
Yeah, this looks like the right direction. Some thoughts.

I think clj-kondo works with `schema.core/defn`, which is similar to `clojure.core.typed/defn`. But that is implemented internally IIRC, unsure if we can leverage that.

The hooks should be put in `typed/clj.runtime/resources`, since typed.clj.runtime is the jar the macros are distributed in.

I'd like to share code between the actual macros and the clj-kondo hooks. In particular, to call `clojure.core.typed.internal/parse-defn*` from the hooks.

**Comment by owenRiddy** [comment]
Hmm. We quickly run up against the limits of clj-kondo's hook implementation.

I had a look at how `schema.core/defn` is implemented in clj-kondo. There is hard-coded support - the implementation lives [here](https://github.com/clj-kondo/clj-kondo/blob/master/src/clj_kondo/impl/schema.clj) and is then called by the [analyzer](https://github.com/clj-kondo/clj-kondo/blob/master/src/clj_kondo/impl/analyzer.clj#L1247). The major issue here, I think, it that hooks (what the PR at hand uses) can rewrite the specific form for linting but the analyzer is a different code path. So in this case, the hook will lint errors for `(t/defn foo [bar :- A] :- A bar)` but calls to `foo` later in the code will claim that the var hasn't been defined because the analyzer isn't clever enough to use the rewritten code. I might go raise a ticket with clj-kondo and see if there is any guidance on what borkdude expects to happen.

I also had a look at sharing the parser code. I suspect there is no elegant way to do that. The hook seems to be run in an environment that is independent of the project. I tried a `:require` in the hook and that didn't help. It may be possible to dodgy something up by copying files into the hook folder at deploy time, but the cure sounds about as bad as the disease.

**Comment by frenchy64** [comment]
Thanks for investigating. I'm fine with copying code, there's a templating system in the repository for generating code that gets committed to the repo. It's used to manage the deps.edn's and pom.xml's.

I will spend some time today investigating the code sharing issue.

**Comment by owenRiddy** [comment]
Cheers. Still not tested but I've made some changes:

1. Resources live in `typed/clj.runtime/resources`.

2. I inquired about how support was added for schema (clj-kondo #2210). Turns out that was just a pre-hook way of doing things and there isn't anything there to copy. What I thought were the reasons were actually bugs in my hook. The code has been improved based on feedback (still light on checks and testing).

I'm going to wait and see what you come up with on code sharing since that will influence what happens next.

**Comment by frenchy64** [comment]
I merged the very basics, I started with `ann-form` which is much simpler than `defn`. https://github.com/typedclojure/typedclojure/commit/1239c81cfee33bf77271339c60f5912a3b98be5d

I'll think some more about code sharing.

**Comment by frenchy64** [comment]
Figured it out! Let's continue the discussion in #123.

**Pull request closed without merging**